frontend/src/metabase-lib/expressions/resolver.js

```diff
@@ -144,13 +144,19 @@ export function resolve(expression, type = "expression", fn = undefined) {
     if (!clause) {
       throw new ResolverError(t`Unknown function ${op}`, expression.node);
     }
-    const { displayName, args, multiple, hasOptions } = clause;
+    const { displayName, args, multiple, hasOptions, validator } = clause;
     if (!isCompatible(type, clause.type)) {
       throw new ResolverError(
         t`Expecting ${type} but found function ${displayName} returning ${clause.type}`,
         expression.node,
       );
     }
+    if (typeof validator === "function") {
+      const validationError = validator.apply(null, operands);
+      if (typeof validationError === "string") {
```

Comment:
This could (or should) really be using an array of strings to signal errors, but for now, there's literally only one error that is handled this way.

Reply:
OTOH, while this catching the invalid `0` as the first parameter through syntax analysis, what happens when the expression is something like `substring([First Name], 4-4, 1)`? What will be the thrown error?

Reply:
The schema is a positive integer, or else any numeric expression. We now clearly indicate that substring is 1-based, and the check for literals is as much as we can really do at this point. Anything else would require evaluating and simplifying numeric expressions on the client and server side and I don't think we want to do that just for this issue.

Reply:
In that case, a situation like `0` as the first parameter should be treated as a run-time error, not a syntactical error (due to the said limitation).

The fact that there is an error message `Expected positive integer but found 0` is then a mere courtesy, and carried out on a best-case basis (as you said, we can't go down the rabbit-hole of constant folding and whatnot).

If that's agreed on, then the best place to surface the said error message will be in `frontend/src/metabase-lib/expressions/diagnostics.js`. **However**, in the interest of getting the first iteration of this PR merged (and possibly tackle or deal with more pressing issues such as compatibility with existing expressions), I think it's fine to have the above error handling **as is** for now, though it will be useful to create a new issue to move the parameter best-case diagnostic to `diagnostics.js`, so that we won't forget it.

Reply:
Ideally, I'd hope this could be driven by the same MBQL schema shared between BE and FE, which would replace both the frontend resolver and diagnostics. I'm not sure if that's part of the metabase-lib idea or not.

Reply:
While I agree with the overall idea, technically it'll likely take more than just a schema to replace the custom expression diagnostics at the front-end, since it involves some work at the lexical level (see #18773 and related PRs for more historical context).